Implement basic battery manager.

kareltucek opened the pull request:
Solves the most pressing issues of #974.

Closes #1206.
Closes #1210.
Closes #1011.
Closes #1214.

What it does:
- implement controls that allow easily setting battery limits
- implement a state machine that enables and disables the charger as needed
- implement a charging mode for stationary use that targets 3.8V, and its macro controls `set battery.chargeLimit { full | optimizeHealth }`
- eliminate the step of battery indicators between powered and unpowered states and various brightness states.
- blink battery-low icon on the oled when battery is low
- implement setLedTxt for uhk80
- fix key chatter on uhk60 when letters are typed while mouse-keys are being used

Furthermore:
- switch zephyr to deferred logging.
- implement a new zephyr logging backend and forward logs to Agent via a new buffer.
- move status buffer to wormhole region and log crashlogs there.

What it does not do:
- adaptively detect 100% voltage
- we do not calibrate according to actual battery health (yet)

kareltucek committed:
Implement basic battery manager.

kareltucek commented:
This isn't tested very well yet.

kareltucek committed:
Fix compilation.

kareltucek committed:
Implement gentle smoothing and add some temporary logs.

kareltucek commented:
Known issues:
- ~~battery presence detection is broken.~~ fixed
- there is still the step caused by whether battery is being drained or not during measurement. (I don't plan to fix this at the moment.)
- ~~i yet plan to implement shutdown mode when battery drops bellow "0". (It looks really bad when battery indicator says 0, but the keyboard still runs happily.~~ implemented, although in a very basic manner.

kareltucek committed:
Fix battery presence detection.

kareltucek committed:
Put the board to sleep when battery is below cutoff threshold.

mondalaci commented:
The code looks very good and well abstracted.

If the charger is disconnected, what policy will you implement for `ChargeRegion_Empty` vs. `ChargeRegion_AlmostEmpty`? It might be worth displaying `Charge me` on the OLED for either or both and not enabling per-key backlight in the first place.

kareltucek commented:
For empty, uhk shuts down (as deep as I can make it).

Almost empty is just an auxiliarry region that makes sure that the keyboard doesnt oscillate between power on and off.

I am not displaying anything at the moment.

mondalaci commented:
Sounds reasonable.

`Charge me` may make sense for `ChargeRegion_AlmostEmpty`, but we'll see what works in practice.

kareltucek commented:
I am not strictly against introducing the warning, but I would prefer to split into a separate issue (I am already writing another code on top of this, am not sure where to render it, and am not even sure we need it)...

kareltucek committed:
Merge remote-tracking branch 'origin/master' into battery_management

kareltucek commented:
I think this is ready for merge.

kareltucek commented:
Actually, not yet. Half will not wake up if restarted 🤦.

kareltucek committed:
Implement power modes and sfjl sleep

kareltucek committed:
Merge remote-tracking branch 'uhk/ble_hid_to_ble_hid' into battery_management

kareltucek committed:
Merge branch 'ble_hid_to_ble_hid' into battery_management

kareltucek committed:
Merge remote-tracking branch 'origin/battery_management' into battery_management

kareltucek committed:
Implement our logging backend.

kareltucek committed:
Protect the wormhole.

kareltucek committed:
Use zephyr logging for bluetooth messages.

kareltucek committed:
Undisable sleep mode.

kareltucek committed:
Optimize key polling when no keys are pressed.

kareltucek committed:
Increase uhk60 noinit segment size to fit logs.

kareltucek committed:
Merge remote-tracking branch 'origin/master' into battery_management

kareltucek committed:
Adjust backlight.

kareltucek committed:
Fix some led display brightness bugs.

kareltucek committed:
Merge remote-tracking branch 'origin/master' into battery_management

mhantsch commented:
Instead of a "charge me" message (and not knowing where to place it), I would flash the battery icon slowly (e.g. 1/sec).

mondalaci commented:
I'd center-align "charge me". A center-aligned battery icon should also work, probably even better.

When running `powerMode sleep`, both halves go to sleep, but the right half displays the colors of the Fn2 layer actions with minimum brightness, and the left half has S and D on with minimum brightness.

S+F or J+L need very little time to activate. I think it'd be useful to request that you press these keys (and no other key) for at least half a second to activate.

I suggest using the `battery.lifeExtenderMode` name instead of `battery.stationaryMode`.

I need explicit instructions on how to test whether backlight oscillation is noticeable during battery voltage measurement. I use a USB and a bridge cable. How long should I observe it? I also believe that the charge level is an important variable, and the backlight oscillation may be more easily noticeable at low battery levels.

mondalaci commented:
Also, red-blinking S+F and J+L would consume less power than white-blinking them, so I'd prefer red.

mondalaci commented:
I may be wrong about the requested half-second activation time. It may not be necessary in practice. Feel free to ignore it; real-world usage will reveal whether it's justified.

mondalaci commented:
`shutDown` should be named `shutdown`, as it's a single word. But more importantly, what's its purpose, as I can't wake up my UHK after shutting it down, except by resetting it? The dim LED issue applies, as with `sleep`.

kareltucek commented:
> When running powerMode sleep, both halves go to sleep, but the right half displays the colors of the Fn2 layer actions with minimum brightness, and the left half has S and D on with minimum brightness.

Noted.

> I suggest using the battery.lifeExtenderMode name instead of battery.stationaryMode.

Stationary mode is I think a common term for this, used for instance with laptop batteries. @mhantsch @pcooke9 (anyone else) opinions?

![2025-04-03-112149_1293x654_scrot](https://github.com/user-attachments/assets/9dac318f-2018-49b8-a499-782f953c32dc)

> shutDown should be named shutdown

Sure!

> S+F or J+L need very little time to activate. I think it'd be useful to request that you press these keys (and no other key) for at least half a second to activate.
> ...
> I may be wrong about the requested half-second activation time. It may not be necessary in practice. Feel free to ignore it; real-world usage will reveal whether it's justified

I don't think there is a risk of accidental wake with current sf/jl, but it means I can increase the polling time 🤔. 

> But more importantly, what's its purpose, as I can't wake up my UHK after shutting it down, except by resetting it? The dim LED issue applies, as with sleep.

It is a mode that UHK enters when the battery runs out. Uhk wakes from it when you connect it to usb (power).  Users are not meant to use it directly.

> I need explicit instructions on how to test whether backlight oscillation is noticeable during battery voltage measurement. I use a USB and a bridge cable. How long should I observe it? I also believe that the charge level is an important variable, and the backlight oscillation may be more easily noticeable at low battery levels.

Well, in `charger.c`, you can drop
```
#define CHARGER_UPDATE_PERIOD 60000
```
to something like 2000.

But in your original report you didn't need to do that and yet your UHK oscillated with a ~1 second-ish period. How you did it I don't know.

I expect the backlight oscillation to be noticeable when the battery is actively charging (i.e., powered, but not full).

mondalaci commented:
I don't insist on the suggested naming, but "stationary mode" doesn't ring a bell to me, so I devised the alternative using Claude. We should use what is most intuitive for most users.

Would you do two consecutive checks for the polling time regarding SF/JL? It's probably the best approach.

The shutdown mode should be a development feature and may be exposed separately as such. However, the sleep and lock modes will be directly exposed as device actions via Agent for easy use, so it may not matter. Please expose the relevant device actions for Robi.

My batteries are at 100%, and I can't notice the oscillation. Let me deplete them and I'll retest.

kareltucek commented:
> The shutdown mode should be a development feature and may be exposed separately as such. However, the sleep and lock modes will be directly exposed as device actions via Agent for easy use, so it may not matter. Please expose the relevant device actions for Robi.

The sleep mode is exposed. 

I am not sure the lock mode should be exposed actually, as it doesn't reduce power consumption.

kareltucek commented:
> My batteries are at 100%, and I can't notice the oscillation. Let me deplete them and I'll retest.

(I think it should be proportional to the current that charges the battery actually, so the lower the battery, the more pronounced it should be.)

mondalaci commented:
I'd also expose the lock mode as a device action. Imagine someone carrying his UHK around without getting disconnected via wireless.

kareltucek committed:
Resolve PR findings.

kareltucek commented:
All suggestions should be addressed except for "charge me" text and the battery mode name. I am not at all convinced by `battery.lifeExtenderMode`. Colleagues suggest "apple mode" and "optimized charging mode". Will leave it as is until more feedback is gathered.

kareltucek committed:
Add a proxy buffer for zephyr logs.

mhantsch commented:
> Also, red-blinking S+F and J+L would consume less power than white-blinking them, so I'd prefer red.

5% of men have red-green colour deficiency. I am one of them. I can hardly see red, or it looks much darker to me than other colours. I don't think red is a good choice.

mondalaci commented:
Thanks, I wouldn't have ever thought about that! I recommend blue then.

pcooke9 commented:
As for the battery mode; I kinda like "Stationary mode" because it makes you think of the use cases. But I think something like "battery protect" is probably more intuitive.

~~The description text for each mode is great the way they are in the image though.~~ 🤦🏻Sorry, hadn't had coffee yet.😂 

EDIT: Maybe better wording would be `battery.preserveHealthMode` or `battery.protectHealthMode`.
(I think Laszlo's "lifeExtender" wording is probably closer in spirit to what the function actually does, but the word "extend" generally refers to adding more to something, or to make something larger/longer, which isn't _technically_ possible in terms of battery life.)

mondalaci commented:
Thanks for your suggestions! I like `battery.preserveHealthMode`.

mhantsch commented:
I'm going to throw my €0.02 in:

```
set battery.chargeLimit 75    // maximum percentage to charge to
set battery.chargeLimit full   // charge to 100%
set battery.chargeLimit optimizeHealth // charge to 80%
set battery.chargeLimit stationaryUse   // charge to 50%
```

kareltucek commented:
@mhantsch thanks, but the feature is set to maintain the lipol storage optimum which is (according to my sources) 3800 mV, which corresponds to roughly 83% according to current percentage computation. It is intentionally anchored to voltage and not to percentage, and intentionally not exposed to the user, as they would tend to leave the battery undercharged, incorrectly believing they are doing a good thing for their battery.

pcooke9 commented:
MAX Concise strikes again!😆
That does look more logical, for sure.

~~Is 50% actually considered optimal for stationary use? I don't think I've ever seen any info on that use case.~~ Karel's comment answered my question before I could ask.🏃‍♂️

kareltucek commented:
Although, I must admit I like the proposed syntax because it is more descriptive as well as more future proof.

I suggest using this syntax. Internally, it would still map to a simple boolean.

```
set battery.chargeLimit full
set battery.chargeLimit optimizeHealth
```

~~@mondalaci ?~~ Edit: implemented this variant. Feel free to protest.

kareltucek committed:
Refine shutdown wake conditions.

kareltucek committed:
Log reason of advertising attempts.

kareltucek committed:
Relax depleted battery mode waking conditions.

kareltucek committed:
Change sfjl colors to blue.

mhantsch commented:
> 3800 mV, which corresponds to roughly 83%

I always thought for long-term storage, it's best to keep LiPos around half filled, i.e. somewhere between 40%-60%. And up to ~80% you still get very little degradation, but benefit from more energy being available in the battery in case you have to go off the charger.

In fact, when I get fresh replacement batteries e.g. for my phone, or my laptop, they arrive about half charged when I take them out of the box. Clearly they come at "best storage level", and instructions are to fully charge them once at the beginning (so the BMS can learn their full capacity).

I understand LiPos age most when kept at less than 20% or higher than 80% for prolonged periods of time.

I can't translate that to voltage; I don't have data for that.

mhantsch commented:
![Lipo-battery-guide-Voltage-vs-capacity-used-percentage](https://github.com/user-attachments/assets/9c970ede-896f-4a80-be58-85201d0dfd06)

https://oscarliang.com/wp-content/uploads/2017/02/Lipo-battery-guide-Voltage-vs-capacity-used-percentage.jpg

kareltucek commented:
There are other issues, like our charging voltage and the percentage calculation. The charger cannot provide more than ~4.1V, which means that for us, 4000mV is 100%.

Rest assured that we charge to 3.8V, whatever the actual percentage is.

kareltucek committed:
Rename battery.stationaryMode to battery.chargeLimit

kareltucek commented:
As far as user experience goes, I feel like 500ms scanning rate and blue blinking is bad. I suggest dropping to 75ms and white blinking.

kareltucek committed:
Fix compilation on uhk60.

kareltucek committed:
Implement generic notification, and use it for "Charge me!" message.

kareltucek committed:
Finetune battery manager settings based on our discharge curves.

kareltucek committed:
Implement voltage correction.

kareltucek committed:
Remove anti-wake timer for automatic shutdown.

kareltucek committed:
Merge branch 'battery_calculator' into battery_management

kareltucek committed:
Compute battery percent based on discharge curves.

kareltucek committed:
Refactor, fill in better default measurements.

kareltucek committed:
Add window average for battery voltage.

kareltucek committed:
Charger fixes.

kareltucek committed:
Blink battery icon when battery is low.

kareltucek commented:
Another chapter of battery management should be finished. I would not start another until this one is tested and merged in master. 

Current state of affairs:
- Charger should be returning consistent battery voltages / percentages. E.g., when plugging the charger in and out, or when turning backlight on or off when running on battery.
- Battery percentage is now computed based on actual charging curves.
- I have raised the cutoff limit to 3.4V after seeing the battery curves, and think we should consider raising it to 3.5V.

We do not adapt to actual battery health at this point - instead we use a pre-measured statistics. Switching to some adaptive or manual calibration will be another step at some point in the future.

At this point I suggest to test and merge this and see if any issues pop up.

kareltucek committed:
Merge remote-tracking branch 'origin/master' into battery_management

kareltucek committed:
Remove forgotten testing code 😅.

kareltucek commented:
Oh, and you can set `#define DEBUG_BATTERY_TESTING true` to test stuff in `debug.h`.

It adds some logs, disables voltage averaging, and shortens battery measurement intervals, so it is possible to observe any quirks much more easily.

kareltucek committed:
Fix average computation.

kareltucek committed:
Implement setLedTxt for uhk80.

kareltucek committed:
DOCS: fill in documentation for new setLedTxt.

kareltucek committed:
Fix warnings.

kareltucek committed:
Make charger start in "charged" state if possible.

kareltucek committed:
Turn on battery averagin by default. (Forgotten code.)

kareltucek committed:
Fix compilation on uhk60, improve logging on uhk80.

kareltucek committed:
Fix dongle compilation.

kareltucek committed:
Adjust logging.

kareltucek committed:
Fix reboot loop on empty but usb-powered battery.

kareltucek committed:
Don't correct 0 voltages.

kareltucek committed:
Change logging, amend battery corrections.

kareltucek committed:
Remove brightness limiting.

kareltucek committed:
Charger fixes and simplifications.

kareltucek committed:
setLedTxt fixes.

kareltucek committed:
Fix setLedTxt dry run validation warnings.

kareltucek committed:
Fix voltage averaging.

kareltucek committed:
Remove some logs.

kareltucek committed:
Logging amends.

kareltucek committed:
Blink low battery states.

kareltucek committed:
Fix voltage averaging algorithm.

kareltucek committed:
Apply new charging icons.

kareltucek committed:
Merge remote-tracking branch 'origin/battery_management' into battery_management

kareltucek committed:
Fix charger automaton start.

kareltucek committed:
Alter battery icons.

kareltucek committed:
Adjust battery spacing.

kareltucek committed:
Remove blinking test.

kareltucek committed:
Fix icon data offsets.

kareltucek committed:
Turn off backlight under certain battery level.

kareltucek committed:
Make backlight threshold wider to prevent oscillations.

kareltucek committed:
Blink battery icon depending on backlight powersaving mode.

kareltucek committed:
Set battery-related values for release.

kareltucek committed:
Inn chargeLimit=full, always charge when less than 100%

kareltucek committed:
Fix power icon update.

When we change whether charging or not, we need to check the charger state again. Simplest way is to set next update to be short.

kareltucek committed:
Set oled brightness to minimum when powersaving.

kareltucek committed:
Fix report resending interactions.

Assume two reports are sent at the same time (say keyboard and mouse),
and one errors (say mouse), the other one succeeds (say keyboard). As
a result of success, keyboard reports are switched and zeroed. When
re-sending, both reports are re-sent, except this time keyboard
interface is re-sending an empty report!

This could also be fixed by calling `mergeReports()` even when
resending. I wonder if it would have been a better approach.

kareltucek committed:
Fix uhk60 wormhole logs

kareltucek committed:
Merge remote-tracking branch 'origin/master' into battery_management

kareltucek committed:
Add charging led brightness mode.

kareltucek committed:
Customize tracing for uhk60.

kareltucek committed:
Revert user config version bump. We don't need it for testing yet.

kareltucek committed:
Add reboot traces, to recognize invalid traces.

kareltucek committed:
Implement action compression
